### ISSUE TITLE
[contributions welcome] sequence parallel training

### ISSUE THREAD

**Issue by rizar** [ServiceNow/PipelineRL#16]
A much needed enhancement for PipelineRL is sequence parallel training. Here's how we can implement it.

1. Move the formation of the batch queue to the `preprocess` stage. Currently every trainer has a thread that reads incoming samples and packs them in microbatches of the right size. That same thread also discards old samples when newer are available. If a group of trainers processes same sequences, they should be working on the same microbatches at all times. We can use Redis to hold a queue of minibatch parts for each training worker, the preprocessor worker will be adding batch parts to these queues as the training workers consume them. 

2. Initialize a process groups for each sequence parallel worker group. We already have an extra torch process group for weight broadcast, we can use the same mechanism to add the sequence parallel groups.

3. Substitute the attention implementation in the model by a copy-paste with DeepSpeed Ulysses in it. That's how veRL does it. Possibly Huggingface plans a better native support of sequence parallel training, worth asking them.

4. Add a check that all workers in each sequence parallel groups are located at the same node. 

That should be it. One should extremely carefully check that the training loss precisely matches the values before this change.

**Comment by rizar** [comment]
will be soon closed by #59
